File: src/test/java/org/geepawhill/contentment/test/PointAssert.java
package org.geepawhill.contentment.test;

import org.assertj.core.api.AbstractAssert;
import org.assertj.core.data.Offset;
import org.assertj.core.util.DoubleComparator;
import org.geepawhill.contentment.geometry.Point;

public class PointAssert extends AbstractAssert<PointAssert, Point>
{
	public PointAssert(Point actual)
	{
		super(actual, PointAssert.class);
	}

	public PointAssert isEqualTo(Object expected)
	{
		if(!(expected instanceof Point)) failWithMessage("Not a point.");
		Point expectedPoint = (Point) expected;
		double offset = 1d;
		DoubleComparator comparator = new DoubleComparator(offset);
		if (comparator.compare(actual.getX(), expectedPoint.getX()) != 0)
			failWithMessage("X Not close enough. Expected: " + expectedPoint.getX() + " within " + offset + " but was: " + actual.getX());
		if (comparator.compare(actual.getY(), expectedPoint.getY()) != 0)
			failWithMessage("y Not close enough. Expected: " + expectedPoint.getY() + " within " + offset + " but was: " + actual.getY());
		return this;
	}

	public PointAssert isGoodEnough(Point expected, Offset<Double> offset)
	{
		DoubleComparator comparator = new DoubleComparator(offset.value);
		if (comparator.compare(actual.getX(), expected.getX()) != 0)
			failWithMessage("X Not close enough. Expected: " + expected.getX() + " within " + offset + " but was: " + actual.getX());
		if (comparator.compare(actual.getY(), expected.getY()) != 0)
			failWithMessage("y Not close enough. Expected: " + expected.getY() + " within " + offset + " but was: " + actual.getY());
		return this;
	}
}

File: src/test/kotlin/test/PointAssert.kt
package test

import org.assertj.core.api.AbstractAssert
import org.assertj.core.data.Offset
import org.assertj.core.util.DoubleComparator
import org.geepawhill.contentment.geometry.Point

class PointAssert(actual: Point) : AbstractAssert<PointAssert, Point>(actual, PointAssert::class.java) {

    override fun isEqualTo(expected: Any): PointAssert {
        if (expected !is Point) failWithMessage("Not a point.")
        val expectedPoint = expected as Point
        val offset = 1.0
        val comparator = DoubleComparator(offset)
        if (comparator.compare(actual.x, expectedPoint.x) != 0)
            failWithMessage("X Not close enough. Expected: " + expectedPoint.x + " within " + offset + " but was: " + actual.x)
        if (comparator.compare(actual.y, expectedPoint.y) != 0)
            failWithMessage("y Not close enough. Expected: " + expectedPoint.y + " within " + offset + " but was: " + actual.y)
        return this
    }

    fun isGoodEnough(expected: Point, offset: Offset<Double>): PointAssert {
        val comparator = DoubleComparator(offset.value)
        if (comparator.compare(actual.x, expected.x) != 0)
            failWithMessage("X Not close enough. Expected: " + expected.x + " within " + offset + " but was: " + actual.x)
        if (comparator.compare(actual.y, expected.y) != 0)
            failWithMessage("y Not close enough. Expected: " + expected.y + " within " + offset + " but was: " + actual.y)
        return this
    }
}
